python/ray/tune/trainable/function_trainable.py

```diff
@@ -243,27 +243,56 @@ def has_new_checkpoint(self):
         return self._fresh_checkpoint
 
     def get_checkpoint(self):
+        # NOTE: This is not the same as `train.get_checkpoint`.
+        # This is used internally by `FunctionTrainable.save_checkpoint`.
+        # `loaded_checkpoint` is the checkpoint accessible by the user.
         self._fresh_checkpoint = False
         return self._last_checkpoint
 
     def _start(self):
         self._last_report_time = time.time()
 
     def report(self, metrics: Dict, *, checkpoint: Optional[Checkpoint] = None) -> None:
+        from ray.train._internal.storage import _use_storage_context
+        from ray.train._internal.checkpoint_manager import _TrainingResult
+        from ray.train.checkpoint import Checkpoint as NewCheckpoint
+
         # TODO(xwjiang): Tons of optimizations.
         self._air_session_has_reported = True
-        if checkpoint:
-            training_iteration = self._get_training_iteration()
-            checkpoint_dir = self.make_checkpoint_dir(step=training_iteration)
-            self.set_checkpoint(checkpoint_dir)
-            checkpoint.to_directory(checkpoint_dir)
-            # TODO(krfricke): Remove this once support is added in Checkpoint.
-            open(os.path.join(checkpoint_dir, ".is_checkpoint"), "a").close()
+
+        # TODO(justinvyu): With a unified session, we'll still run into this doubled
```

Comment:
Can you elaborate on why we'd still have a doubled upload problem?

Reply:
It's just because we call `train.report` internally to pipe data from Train -> Tune:

* The first `train.report` is called by the user in their worker code: `train.report(Checkpoint.from_directory(local_dir)`
    * We upload the checkpoint and send `Checkpoint(fs=storage_filesystem, path=".../checkpoint_0000x")` from each worker to the coordinator (trainable).
* The second `train.report` is now called in `DataParallelTrainer`: `train.report(..., Checkpoint(fs=storage_filesystem, path=".../checkpoint_0000x"))`.
    * All workers have sent back this checkpoint, and now we need to report it to Tune for book-keeping.
    * We shouldn't try to upload anything on this 2nd call.

Reply:
Why is there a second report that DataParallelTrainer calls? I thought train.report() is only ever called by user code?

Reply:
Is this only for multi-rank use case (worker vs driver?)

Reply:
See here: https://github.com/ray-project/ray/pull/37888/files#diff-1c8156765a66e01c9b2bfa118efd58fa3a2551380d33c2898d4289da37cd70e3R431-R443

For Tune Trainables, there's only 2 layers:
```
driver <--(user calls train.report)-- Trainable
```

For `DataParallelTrainer`, there's 3 layers:

```
driver <--(we call train.report)-- Trainable <--(user calls train.report)-- Train worker
```

We use the same "session report" abstraction twice for the `DataParallelTrainer`. 

When Tune and Train were separate before AIR, there was only 2 layers:

```
Train driver <--(train.report)-- Train worker
```

But then, you'd need to re-implement the whole control loop, and it `Tuner(Trainer)` is not so easy to implement, which is why the execution backend got switched to Tune --> 3 layers.